### PR TITLE
Migrate `x86_64-fortanix-unknown-sgx-lvi` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

The final Makefile! Every Makefile test is now claimed.

This is difficult to test due to the uncommon architecture it is specific to. I don't think it is in the CI (I didn't find it in `jobs.yml`, but if there is a way to test it, please do.

Locally, on Linux, it compiles and panics at the `llvm_filecheck` part (if I replace the `x86_64-fortanix-unknown-sgx` with `x86_64-unknown-linux-gnu`, of course), which is expected.

For this reason, the Makefile and associated script have been kept, but with a leading underscore.

### CODE REVIEW

**Comment by jieyouxu** [comment]
I'm going to be real with you, I have absolutely no clue about sgx. So I'm going to summon our sgx target maintainers after I take a look and try to guess what the test is trying to test.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (workingjubilee) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:43]:
```suggestion
    check("cmake_plus_one_c", "cmake_plus_one_c.checks");
    check("cmake_plus_one_c_asm", "cmake_plus_one_c_asm.checks");
    check("cmake_plus_one_c_global_asm", "cmake_plus_one_c_global_asm.checks");
    check("cmake_plus_one_cxx", "cmake_plus_one_cxx.checks");
    check("cmake_plus_one_cxx_asm", "cmake_plus_one_cxx_asm.checks");
```

Comment (workingjubilee) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:43]: This is about verifying that LLVM cross-compiles SGX code with [load value injection] defenses.

It should be possible to make it work after building the cross-compile target.

**Comment by workingjubilee** [comment]
This is about verifying that LLVM cross-compiles SGX code with [load value injection](https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/technical-documentation/load-value-injection.html) defenses.

It should be possible to make it work after building the cross-compile target.

**Comment by workingjubilee** [comment]
This target is tier 2 so we should be building the cross-compilation artifacts every time anyways.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (Oneirical) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs]: Thank you!

**Comment by workingjubilee** [comment]
So: per policy, tests don't have to pass on SGX, a tier 2 target, in order to merge this. As this is an SGX-only test, we can just merge this through.

I do know, however, that the SGX maintainers do in fact run the CI regularly and add ignore-sgx, etc. and are reasonably responsive, usually.

@jethrogb @raoulstrackx @mzohreva Can we get your review?

**Comment by raoulstrackx** [comment]
Thanks for tagging @workingjubilee ! I have a day off today, but I'll try to look at it on Monday.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:48]: Suggestion: for unknowledgeable future readers (like me), it would be very helpful if there are brief summaries of the steps we are taking here to perform the check, i.e. what the steps are doing (and even better, *why*).

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:63]: Suggestion: could we try to figure out *why* this is special-cased?

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:89]: Suggestion: this link should not be wrapped, please add a `ignore-tidy-linelength` if you have to.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:45]: EDIT: I see there's a link to <https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/technical-documentation/load-value-injection.html> below, we should hoist that link (if not duplicate it) to the top-level comment so it's immediately available.

Based on that document (reproduced partially below, and trimmed, but it leaves out some other important context):

> ## Overview of Load Value Injection
>
> On some processors, faulting or assisting[^1] load operations may transiently receive data from a microarchitectural buffer[^2].
> If an adversary can cause a specified victim load to fault, assist, or abort, the adversary may be able to select the data to have forwarded to dependent operations by the faulting/assisting/aborting load. For certain code sequences, those dependent operations may create a covert channel with data of interest to the adversary. The adversary may then be able to infer the data's value through analyzing the covert channel. This transient execution attack[^3] is called load value injection (LVI) and is an example of a cross-domain transient execution attack.

[^1]: Assists are conditions that are handled internally by the processor and thus do not require software involvement. While both faults and assists may cause the results of a μop to be discarded, assists restart and complete the instruction without needing software involvement, whereas faults do need software involvement (for example, an exception handler). For example, setting the Dirty bit in a page table entry may be done using an assist.
[^2]: Some processors may transiently receive incorrect data from the store buffer, load ports, fill buffer, or L1D cache. For more details, please see [Intel Analysis of Microarchitectural Data Sampling](https://www.intel.com/content/www/us/en/develop/articles/software-security-guidance/technical-documentation//intel-analysis-microarchitectural-data-sampling.html) and [Intel Analysis of L1 Terminal Fault](https://www.intel.com/content/www/us/en/develop/articles/software-security-guidance/technical-documentation/intel-analysis-l1-terminal-fault.html).
[^3]: Refer to [Refined Speculative Execution Terminology](https://www.intel.com/content/www/us/en/develop/articles/software-security-guidance/best-practices/refined-speculative-execution-terminology.html).

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:3]: Question (to sgx maintainers): is this statement accurate, or is the test demonstrating a particular kind of hardening or mitigations for a particular kind of LVI attack?

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:45]: I read this document, but as a non-security researcher, I failed to come up with a good TL;DR for a future reader not knowledgeable about LVI mitigations. So maybe it's simply better to just leave the details to the document, but try to describe *why* do the checks we perform here is sufficient to demonstrate the specific kind of hardening intended for a specific kind of LVI attack method/step.

(Please help me replace "specific" with... the actual "specific" terminology if possible)

**Review by jieyouxu — COMMENTED** [review]
Thanks for the port, I have *several* questions (that we will need help from the sgx maintainers).

**Comment by jieyouxu** [comment]
> Thanks for tagging @workingjubilee ! I have a day off today, but I'll try to look at it on Monday.

Thank you very much for the help!

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:3]: (Also, as mentioned in another review comment, do we know which particular kinds here, if applicable?)

Comment (raoulstrackx) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:6]: The tests don't differentiate between different types of LVI attacks

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (raoulstrackx) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:66]: The script does this:
```
    func="$(${objdump} --syms --demangle "${enclave}" | \
            grep --only-matching -E "[[:blank:]]+${func_re}\$" | \
            sed -e 's/^[[:space:]]*//' )"
```
This is needed. Otherwise you get output like:
```                                                                                                                                                                                                                                            
enclave/target/x86_64-fortanix-unknown-sgx/debug/enclave:       file format elf64-x86-64                                                                                                                                                    
                                                                                                                                                                                                                                            
SYMBOL TABLE:                                                                                                                                                                                                                               
0000000000000000 l    df *ABS*  0000000000000000 30q2fp73pc99bmzfe1chpqhcp                                                                                                                                                                  
0000000000012600 l     F .text  000000000000001e .hidden _ZN3std3sys9backtrace28__rust_begin_short_backtrace17hb9bbcbf22e1d0f53E                                                                                                            
0000000000012620 l     F .text  0000000000000006 .hidden __llvm_lvi_thunk_r11 
...
000000000000000c l *ABS* 0000000000000000 tcsls_user_mxcsr                                                                                                                                                                     
0000000000000010 l *ABS* 0000000000000000 tcsls_last_rsp                                                                                                                                                                       
0000000000000018 l *ABS* 0000000000000000 tcsls_panic_last_rsp                                                                                                                                                                 
0000000000000020 l *ABS* 0000000000000000 tcsls_debug_panic_buf_ptr                                                                                                                                                            
0000000000000028 l *ABS* 0000000000000000 tcsls_user_rsp                                                                                                                                                                       
0000000000000030 l *ABS* 0000000000000000 tcsls_user_retip    # <--- false positive on this ret
...
Disassembly of section .text:                                                                                                                                                                                                               
                                                                                                                                                                                                                                            
0000000000031a28 <__libunwind_Registers_x86_64_jumpto>:                                                                                                                                                                                     
   31a28: 48 8b 47 38                   movq    0x38(%rdi), %rax                                                                                                                                                                            
   31a2c: 0f ae e8                      lfence                                                                                                                                                                                              
   31a2f: 48 83 e8 10                   subq    $0x10, %rax                                                                                                                                                                                 
   31a33: 48 89 47 38                   movq    %rax, 0x38(%rdi)                                                                                                                                                                            
   31a37: 48 8b 5f 20                   movq    0x20(%rdi), %rbx                                                                                                                                                                            
   31a3b: 0f ae e8                      lfence   
```
and that gives false positives when checking that `ret` is not present

**Review by raoulstrackx — CHANGES_REQUESTED** [review]
At the moment this PR breaks CI. See comments. Can you please check it? You shouldn't have to have an SGX machine as the enclaves don't need to be executed.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (Oneirical) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:66]: In the original script:

```sh
    func="$(${objdump} --syms --demangle "${enclave}" | \
            grep --only-matching -E "[[:blank:]]+${func_re}\$" | \
            sed -e 's/^[[:space:]]*//' )"
    ${objdump} --disassemble-symbols="${func}" --demangle \
      "${enclave}" > "${asm}"
```

The first objdump seeks all `func_re` instances with a leading space, then trims away the space. The second objdump has no further filtering or normalization.

In the Rust code:

```rs
    let dump = llvm_objdump()
        .input("enclave/target/x86_64-fortanix-unknown-sgx/debug/enclave")
        .args(&["--syms", "--demangle"])
        .run()
        .stdout_utf8();
    let re = regex::Regex::new(&format!("[[:blank:]]+{func_re}")).unwrap();
    let func = re.find_iter(&dump).map(|m| m.as_str().trim()).collect::<Vec<&str>>().join(",");
    let dump = llvm_objdump()
        .input("enclave/target/x86_64-fortanix-unknown-sgx/debug/enclave")
        .args(&["--syms", &format!("--disassemble-symbols={func}")])
        .run()
        .stdout_utf8();
```

The regex pattern is `[[:blank:]]+{func_re}`, which should avoid detecting the `tcsls_user_retip`, and then the `.trim()` replaces the `sed` of the original script. As for the second objdump, it has no filtering or normalizations like the original script.

Let me know if I missed something.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (Oneirical) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:74]: There are two calls of `llvm_filecheck` for each check, except in 3 specific cases:

```sh
    if [ "${func_re}" != "rust_plus_one_global_asm" ] &&
         [ "${func_re}" != "cmake_plus_one_c_global_asm" ] &&
         [ "${func_re}" != "cmake_plus_one_cxx_global_asm" ]; then
        # The assembler cannot avoid explicit `ret` instructions. Sequences
        # of `shlq $0x0, (%rsp); lfence; retq` are used instead.
        # https://www.intel.com/content/www/us/en/developer/articles/technical/
        #     software-security-guidance/technical-documentation/load-value-injection.html
        ${filecheck} --implicit-check-not ret --input-file "${asm}" "${checks}"
    fi
```

This part of the script was meant to handle the "with or without frame pointers" edge case. If there are no frame pointers, the `--implicit-check-not ret` filecheck still needs to run, with the new `without_frame_pointers` check.

I see now that I can modify the `checks` variable to make this part more readable.

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (Oneirical) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs]: That was my bad. I am very sorry if it took you time to find this stealthy mistake!

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs:58]: Thank you, that is very helpful context!

**Comment by raoulstrackx** [comment]
There are other issues as well. Please apply:
```
diff --git a/tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs b/tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs                                                                                                                                                                                              
index 84c292d019d..506b9424d39 100644                                                                                                                                                                                                                                                                                       
--- a/tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs                                                                                                                                                                                                                                                               
+++ b/tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs                                                                                                                                                                                                                                                               
@@ -41,13 +41,13 @@ fn main() {                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                            
     check("std::io::stdio::_print::[[:alnum:]]+", "print.with_frame_pointers.checks");                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                            
-    check("st_plus_one_global_asm", "rust_plus_one_global_asm.checks");                                                                                                                                                                                                                                                    
+    check("rust_plus_one_global_asm", "rust_plus_one_global_asm.checks");                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                            
-    check("_plus_one_c", "cc_plus_one_c.checks");                                                                                                                                                                                                                                                                          
-    check("_plus_one_c_asm", "cc_plus_one_c_asm.checks");                                                                                                                                                                                                                                                                  
-    check("_plus_one_cxx", "cc_plus_one_cxx.checks");                                                                                                                                                                                                                                                                      
-    check("_plus_one_cxx_asm", "cc_plus_one_cxx_asm.checks");                                                                                                                                                                                                                                                              
-    check("_plus_one_asm", "cc_plus_one_asm.checks");                                                                                                                                                                                                                                                                      
+    check("cc_plus_one_c", "cc_plus_one_c.checks");                                                                                                                                                                                                                                                                        
+    check("cc_plus_one_c_asm", "cc_plus_one_c_asm.checks");                                                                                                                                                                                                                                                                
+    check("cc_plus_one_cxx", "cc_plus_one_cxx.checks");                                                                                                                                                                                                                                                                    
+    check("cc_plus_one_cxx_asm", "cc_plus_one_cxx_asm.checks");                                                                                                                                                                                                                                                            
+    check("cc_plus_one_asm", "cc_plus_one_asm.checks");                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                            
     check("cmake_plus_one_c", "cmake_plus_one_c.checks");                                                                                                                                                                                                                                                                  
     check("cmake_plus_one_c_asm", "cmake_plus_one_c_asm.checks");                                                                                                                                                                                                                                                          
@@ -66,9 +66,11 @@ fn check(func_re: &str, mut checks: &str) {                                                                                                                                                                                                                                                              
         .stdout_utf8();                                                                                                                                                                                                                                                                                                    
     let re = regex::Regex::new(&format!("[[:blank:]]+{func_re}")).unwrap();                                                                                                                                                                                                                                                
     let func = re.find_iter(&dump).map(|m| m.as_str().trim()).collect::<Vec<&str>>().join(",");                                                                                                                                                                                                                            
+    assert!(!func.is_empty());                                                                                                                                                                                                                                                                                             
+                                                                                                                                                                                                                                                                                                                           
     let dump = llvm_objdump()                                                                                                                                                                                                                                                                                              
         .input("enclave/target/x86_64-fortanix-unknown-sgx/debug/enclave")                                                                                                                                                                                                                                                 
-        .arg(&format!("--disassemble-symbols={func}"))                                                                                                                                                                                                                                                                     
+        .args(&["--demangle", &format!("--disassemble-symbols={func}")])                                                                                                                                                                                                                                                   
         .run()                                                                                                                                                                                                                                                                                                             
         .stdout_utf8();                                                                                                                                                                                                                                                                                                    
     let dump = dump.as_bytes();
```

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/_Makefile
Comment (raoulstrackx) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/_Makefile:2]: Can you fix this fixme before merging?

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/_script.sh
Comment (raoulstrackx) [inline tests/run-make/x86_64-fortanix-unknown-sgx-lvi/_script.sh:3]: ... and this one as well.

**Review by raoulstrackx — APPROVED** [review]
Now the tests all pass. :)

**Comment by Oneirical** [comment]
> Now the tests all pass. :)

I deleted both `Makefile` and `script.sh` as they are no longer needed, let me know if you'd like to keep them.

Thank you so much for all your help!! This wouldn't have been possible without a fortanix-rust maintainer.

**Comment by jieyouxu** [comment]
> Now the tests all pass. :)

@raoulstrackx thank you so much for helping to review and test 💚

File: tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs
Comment (jieyouxu) [reply tests/run-make/x86_64-fortanix-unknown-sgx-lvi/rmake.rs]: Yes, that is very helpful. Thank you!

**Comment by jieyouxu** [comment]
> I deleted both `Makefile` and `script.sh` as they are no longer needed, let me know if you'd like to keep them.

They're still in the change history, and now the rmake.rs version reimplements them, so we shouldn't keep them.

**Review by jieyouxu — APPROVED** [review]
Thanks, this LGTM as well. Feel free to r=me after PR CI is green.

**Pull request merged**
